packages/react-core/src/Authenticator/hooks/useAuthenticatorProps/utils.ts

```diff
@@ -0,0 +1,81 @@
+import { AuthenticatorRoute } from '@aws-amplify/ui';
+
+import {
+  AuthenticatorMachineContext,
+  AuthenticatorMachineContextKey,
+  AuthenticatorRouteComponentKey,
+} from '../types';
+
+import { UseAuthenticatorSelector } from '../useAuthenticator';
+import { isComponentRouteKey } from '../utils';
+
+import { ROUTE_SELECTOR_KEYS } from './constants';
+import {
+  RoutePropsResolvers,
+  RouteSelectorProps,
+  TranslateHandlerProps,
+  UseAuthenticatorRouteProps,
+} from './types';
+
+// only select `route` from machine context
+const routeOnlySelector: UseAuthenticatorSelector = ({ route }) => [route];
+
+const createSelector =
+  (selectorKeys: AuthenticatorMachineContextKey[]): UseAuthenticatorSelector =>
+  (context) =>
+    // always include `route` in selector
+    [...selectorKeys.map((key) => context[key]), context.route];
+
+export const getRouteMachineSelector = (
+  route: AuthenticatorRoute
+): UseAuthenticatorSelector =>
+  isComponentRouteKey(route)
+    ? createSelector(ROUTE_SELECTOR_KEYS[route])
+    : routeOnlySelector;
+
+function translateHandlerProps<Props extends TranslateHandlerProps>({
+  submitForm: handleSubmit,
+  updateBlur: handleBlur,
+  updateForm: handleChange,
+  ...rest
+}: Props) {
+  return { ...rest, handleBlur, handleChange, handleSubmit };
+}
+
+const PROPS_RESOLVERS: RoutePropsResolvers = {
+  confirmResetPassword: translateHandlerProps,
+  confirmSignIn: (props) => ({
+    ...translateHandlerProps(props),
+    challengeName: props.user.challengeName!,
```

Comment:
Would love to see a note for current and future maintainers explaining the confidence behind using the non-null assertion operator.

Reply:
The issue here isn't so much the non-null assertion but is the way that the state machine and Amplify JS are typed but i don't disagree with you in terms of adding a comment.
```suggestion
    // `challengName` is always provided on `user` for the `confirmSignIn` route
    challengeName: props.user.challengeName!,
```